### PR TITLE
NMS-7404: Create a detector for XMP

### PR DESCRIPTION
**Pull request by rdkgit**
XmpDetector and associated config files for adding detector to ONMS xmp protocol jar.

JIRA: http://issues.opennms.org/browse/NMS-7404
Reference to XMP: http://www.opennms.org/wiki/XMP

Todo:
- [x] OCA signed
- [ ] Code review
- [ ] Pass unit tests
- [ ] Documentation
- [ ] Merge to develop and close JIRA issue

### CODE REVIEW

**Comment by indigo423** [comment]
@rdkgit Thanks seems good, I've cross referenced the JIRA issue and the pull request and changed the issue title a little bit more generic. If we merge the pull request the NMS-7404 is also mentioned in the comment and makes it easy to find commits related to this issue.

I found a wiki page about XMP and added this reference as well. We started documenting all the monitors in OpenNMS and I found the XMP monitor is still an open issue (http://issues.opennms.org/browse/NMS-6702). I can start with this one and with your knowledge we should use this opportunity to get some docs for the monitor and the related detector? What do you think?

**Comment by rdkgit** [comment]
Hi!

>    I found a wiki page about XMP and added this reference as well. We
>    started documenting all the monitors in OpenNMS and I found the XMP
>    monitor is still an open issue
>    ([2]http://issues.opennms.org/browse/NMS-6702). I can start with this
>    one and with your knowledge we should use this opportunity to get some
>    docs for the monitor and the related detector? What do you think?

I will take a look at this one and handle it.

Thanks,

Bobby

**Comment by roskens** [comment]
The changes to files in opennms-provisiond seem unneeded?

**Comment by rdkgit** [comment]
Hi!

>    The changes to files in opennms-provisiond seem unneeded?

Yes, except for one change that has already been put in.

The needed change was adding a check for empty string vs. null string
object in IpInterfaceScan.

Other changes I made were merely to make the code easier for me to
read.

Only other changes to really submit were the XmpDetector related code
-- XmpDetector.java, detectors.xml, and default-foreign-source.xml and
pom.xml

Thanks,

Bobby

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java [hunk 1]
```diff
@@ -179,6 +179,7 @@ public void handleException(final Throwable t) {
     }
 
     Runnable runDetector(final SyncServiceDetector detector, final Callback<Boolean> cb) {
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java]: unnecessary change, should be removed

Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java [hunk 2]
```diff
@@ -213,10 +214,12 @@ Task createDetectorTask(final BatchTask currentPhase, final ServiceDetector dete
     }
 
     private Task createAsyncDetectorTask(final BatchTask currentPhase, final AsyncServiceDetector asyncDetector) {
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java [hunk 3]
```diff
@@ -213,10 +214,12 @@ Task createDetectorTask(final BatchTask currentPhase, final ServiceDetector dete
...
         return currentPhase.getCoordinator().createTask(currentPhase, runDetector(asyncDetector), servicePersister(currentPhase, asyncDetector.getServiceName()));
     }
 
     private Task createSyncDetectorTask(final BatchTask currentPhase, final SyncServiceDetector syncDetector) {
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java [hunk 4]
```diff
@@ -227,8 +230,12 @@ public void run(final BatchTask currentPhase) {
 
         LOG.info("Detecting services for node {}/{} on address {}: found {} detectors", getNodeId(), getForeignSource(), str(getAddress()), detectors.size());
 
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java [hunk 5]
```diff
@@ -227,8 +230,12 @@ public void run(final BatchTask currentPhase) {
...
         for (final ServiceDetector detector : detectors) {
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java [hunk 6]
```diff
@@ -227,8 +230,12 @@ public void run(final BatchTask currentPhase) {
...
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java [hunk 7]
```diff
@@ -227,8 +230,12 @@ public void run(final BatchTask currentPhase) {
 
         LOG.info("Detecting services for node {}/{} on address {}: found {} detectors", getNodeId(), getForeignSource(), str(getAddress()), detectors.size());
 
+
         for (final ServiceDetector detector : detectors) {
+
+	    
             if (shouldDetect(detector)) {
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/IpInterfaceScan.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java [hunk 8]
```diff
@@ -1169,16 +1169,23 @@ private OnmsIpInterface saveOrUpdate(final OnmsIpInterface iface) {
...
+	    aList = new ArrayList<ServiceDetector>(m_pluginRegistry.getAllPlugins(ServiceDetector.class));
+
+	    return aList;
         }
 
         final List<ServiceDetector> detectors = new ArrayList<ServiceDetector>(detectorConfigs.size());
+
         for(final PluginConfig detectorConfig : detectorConfigs) {
+
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java [hunk 9]
```diff
@@ -1169,16 +1169,23 @@ private OnmsIpInterface saveOrUpdate(final OnmsIpInterface iface) {
...
+	    ArrayList aList;
+	    aList = new ArrayList<ServiceDetector>(m_pluginRegistry.getAllPlugins(ServiceDetector.class));
+
+	    return aList;
         }
 
         final List<ServiceDetector> detectors = new ArrayList<ServiceDetector>(detectorConfigs.size());
+
         for(final PluginConfig detectorConfig : detectorConfigs) {
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java [hunk 10]
```diff
@@ -1169,16 +1169,23 @@ private OnmsIpInterface saveOrUpdate(final OnmsIpInterface iface) {
...
         if (detectorConfigs == null) {
-            return new ArrayList<ServiceDetector>(m_pluginRegistry.getAllPlugins(ServiceDetector.class));
+	    ArrayList aList;
+	    aList = new ArrayList<ServiceDetector>(m_pluginRegistry.getAllPlugins(ServiceDetector.class));
+
+	    return aList;
         }
 
         final List<ServiceDetector> detectors = new ArrayList<ServiceDetector>(detectorConfigs.size());
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java [hunk 11]
```diff
@@ -1169,16 +1169,23 @@ private OnmsIpInterface saveOrUpdate(final OnmsIpInterface iface) {
...
         final ForeignSource foreignSource = m_foreignSourceRepository.getForeignSource(foreignSourceName);
         assertNotNull(foreignSource, "Expected a foreignSource with name %s", foreignSourceName);
 
         final List<PluginConfig> detectorConfigs = foreignSource.getDetectors();
         if (detectorConfigs == null) {
-            return new ArrayList<ServiceDetector>(m_pluginRegistry.getAllPlugins(ServiceDetector.class));
+	    ArrayList aList;
+	    aList = new ArrayList<ServiceDetector>(m_pluginRegistry.getAllPlugins(ServiceDetector.class));
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java [hunk 12]
```diff
@@ -1169,16 +1169,23 @@ private OnmsIpInterface saveOrUpdate(final OnmsIpInterface iface) {
     /** {@inheritDoc} */
     @Override
     public List<ServiceDetector> getDetectorsForForeignSource(final String foreignSourceName) {
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultProvisionService.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java [hunk 13]
```diff
@@ -154,11 +155,19 @@ private static void error(Throwable cause, String format, Object... args) {
...
+
         T bean = beans.get(beanName);
-        if (bean != null) debug("Found bean {} with name {} of type {}", bean, beanName, pluginClass);
+
+        if (bean != null) {
+	    debug("Found bean {} with name {} of type {}", bean, beanName, pluginClass);
+	}
+	else {
+            debug("failed to find bean {} with name {} of type {}", bean, beanName, pluginClass);
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java [hunk 14]
```diff
@@ -154,11 +155,19 @@ private static void error(Throwable cause, String format, Object... args) {
     }
     
     private <T> T beanWithNameOfType(String beanName, Class<T> pluginClass) {
+
         Map<String, T> beans = beansOfType(pluginClass);
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java [hunk 15]
```diff
@@ -154,11 +155,19 @@ private static void error(Throwable cause, String format, Object... args) {
     }
     
     private <T> T beanWithNameOfType(String beanName, Class<T> pluginClass) {
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java [hunk 16]
```diff
@@ -131,13 +131,14 @@ private static void error(Throwable cause, String format, Object... args) {
     /** {@inheritDoc} */
     @Override
     public <T> T getPluginInstance(Class<T> pluginClass, PluginConfig pluginConfig) {
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java [hunk 17]
```diff
@@ -131,13 +131,14 @@ private static void error(Throwable cause, String format, Object... args) {
...
         T pluginInstance = beanWithNameOfType(pluginConfig.getPluginClass(), pluginClass);
+
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java]: unnecessary change, should be removed

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java [hunk 18]
```diff
@@ -154,11 +155,19 @@ private static void error(Throwable cause, String format, Object... args) {
...
     private <T> T beanWithNameOfType(String beanName, Class<T> pluginClass) {
+
         Map<String, T> beans = beansOfType(pluginClass);
+
         T bean = beans.get(beanName);
-        if (bean != null) debug("Found bean {} with name {} of type {}", bean, beanName, pluginClass);
+
+        if (bean != null) {
+	    debug("Found bean {} with name {} of type {}", bean, beanName, pluginClass);
+	}
```
Comment (indigo423) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/DefaultPluginRegistry.java]: Code formatting seems strange

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 19]
```diff
@@ -0,0 +1,254 @@
+/*
```
Comment (jeffgdotorg) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Let's get the standard license / copyright copy in here for consistency.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+
+package org.opennms.protocols.xmp.detector;
+
+import java.net.InetAddress;
+import java.util.Date;
+import org.opennms.netmgt.provision.support.BasicDetector;
+import org.opennms.netmgt.provision.support.AbstractDetector;
+import org.opennms.netmgt.provision.ServiceDetector;
+import org.opennms.netmgt.provision.SyncServiceDetector;
+import org.krupczak.xmp.*;
```
Comment (jeffgdotorg) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: I think star-imports are discouraged by the ArsDigita code style guidelines that we more or less follow. Eclipse will clean this up with a Ctrl+Shift+O (organize imports).

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+	// as provisiond's ip range matching functionality
+	// will not correctly match if 0-len instead of null
+	
+	//m_ipMatch = new String("");
+	m_ipMatch = null;  
+
+	createTimeDate = new Date();
+
+	if (LOG == null) {
+	   System.out.println("XmpDetector created, but null LOG");
```
Comment (jeffgdotorg) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: KILL ALL PRINTLN()!!! :)

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 20 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Hi!

How do I do that in emacs?

Bobby

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 21 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Hi!

Printf is the only way to debug!

Bobby

Comment (indigo423) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: If the logger returned by the LoggerFactory is ever be null, it should be considered a bug in slf4j, which doesn't justify the system.out.println. I think this is something what Jeff means, right?

**Comment by agalue** [comment]
Hello,

I encourage you to check http://issues.opennms.org/browse/NMS-6829. Over there you'll find all the information and the code changes behind ipMatch.

All the detector implementations extends AbstractDetector except your XmpDetector, and AbstractDetector contains the correct implementation for ipMatch.

To be consistent with the rest of the detector's implementations I strongly suggest to use AbstractDetector, either through SyncAbstractDetector, or AsyncAbstractDetector, or any class that extends one of the those classes, or AbstractDetector itself.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 21 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Hi!

I'm OK with however you guys want to mangle the submission.  When I
was originally writing the detector, I seem to recall seeing some
other detector issue a simlilar statement.

Bobby

**Comment by rdkgit** [comment]
Hi!

>    I encourage you to check [1]http://issues.opennms.org/browse/NMS-6829.
>    Over there you'll find all the information and the code changes behind
>    ipMatch.
> 
>    All the detector implementations extends AbstractDetector except your
>    XmpDetector, and AbstractDetector contains the correct implementation
>    for ipMatch.

Looking through the code for AbstractDetector, I don't see any
m_ipMatch initialization statements thus there was no indication as to
whether to initialize to null or to zero-length string.

>    To be consistent with the rest of the detector's implementations I
>    strongly suggest to use AbstractDetector, either through
>    SyncAbstractDetector, or AsyncAbstractDetector, or any class that
>    extends one of the those classes, or AbstractDetector itself.

Isn't this correct?

public class XmpDetector implements SyncServiceDetector

Bobby

**Comment by agalue** [comment]
Bod, it is correct, but that means, you are starting from scratch instead of reusing existing functionality. If you implement the interface, you must handle ipMatch and everything inside SyncAbstractDetector by yourself which doesn't seem to be right. I mean, if something is changed on the core logic of AbstractDetector or SyncAbstractDetector, that means, your detector must be reimplemented ;)

When I added the ipMatch feature, I never touched a single detector implementation. I've added the change I need to AbstractDetector and because everything extends directly or indirectly that class, all the detectors implementations are going to use the functionality.

Makes sense ?

**Comment by agalue** [comment]
In other words, if you use AbstractDetector or SyncAbstractDetector, there is no need to worry or know that ipMatch exist, and how it is initialized.

**Comment by rdkgit** [comment]
Hi!

>    In other words, if you use AbstractDetector or SyncAbstractDetector,
>    there is no need to worry or know that ipMatch exist, and how it is
>    initialized.

SyncServiceDetector is an interface and thus provides no inheritance
of ipMatch functionality and neither provide any semantic information
on ipMatch (e.g. null object or empty string).  Oh well, I figured it
out.

Thanks,

Bobby

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+/**
+ * XmpDetector class
+ * @author rdk <rdk@krupczak.org>
+ * @version $Id: $
+ *
+ */
+@Scope("prototype")
+public class XmpDetector implements SyncServiceDetector
+{
+    // class variables
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Please do not add '// class variables' or '// instance variables'.
The meaning is clear due to the fact that it is static.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+    SocketOpts sockopts;
+    XmpConfig protoConfig;
+    int xmpPort;
+    int xmpTimeout;
+    int xmpRetries;
+    String xmpAuthenUser;
+    String xmpServiceName; 
+    String m_ipMatch;
+    Date createTimeDate;
+
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Please use visibility according to your needs. 
Which means by default everything should be `private`.

Please also use one schema to name the private members of your class. Either prefixed with `m_` or without.
Please do not mix.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+        sockopts = new SocketOpts();
+
+        xmpServiceName = new String(serviceName);
+
+	// very important to set to null, not 0-len string
+	// as provisiond's ip range matching functionality
+	// will not correctly match if 0-len instead of null
+	
+	//m_ipMatch = new String("");
+	m_ipMatch = null;
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Please remove this line. The variable is already initialized with `null`.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+     * @param port an int specifying tcp port number
+     */
+    
+    public XmpDetector(String serviceName, int port)
+    {
+	//super(serviceName,port,XMP_DEFAULT_TIMEOUT,XMP_DEFAULT_RETRIES);
+	
+        // set default config
+        xmpPort = port;
+        xmpAuthenUser = new String("xmpUser");
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Please create a static member for the default authenticated user. 
Please do not create a String using `new String("...")` Simply use `"..."`.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+
+    public String getServiceName() {
+	LOG.debug("XmpDetector: getServiceName");
+	return xmpServiceName;
+    }
+  
+    public void setServiceName(String newServiceName) 
+    {
+	LOG.debug("XmpDetector: setServiceName to "+newServiceName);
+    	xmpServiceName = new String(newServiceName);
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Please use `xmpServiceName = newServiceName`.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 27]
```diff
@@ -0,0 +1,254 @@
...
+               xmpPort = protoConfig.getPort();
+            if (protoConfig.hasTimeout())
+               xmpTimeout = protoConfig.getTimeout();
+            if (protoConfig.getAuthenUser() != null)
+               xmpAuthenUser = protoConfig.getAuthenUser();
+
+            sockopts.setConnectTimeout(xmpTimeout);	    
+
+        } catch (Throwable e) {
+            if (LOG != null)
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Please do not catch Throwable, unless it is really necessary. 
Please also handle the exception accordingly --> NO System.out.
Please also do not check if LOG is null. LOG should never be null.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 28]
```diff
@@ -0,0 +1,254 @@
...
+	// dispose of anything like sessions, etc.
+        // no need to dispose SocketOpts
+    }
+
+    @Override
+    public final boolean isServiceDetected(InetAddress address)
+    {
+        XmpSession aSession;
+        XmpMessage aReply;
+        XmpVar[] vars,replyVars;
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Please do not create Multiple variables at once, use
`XmpVar[] vars; `
`XmpVar[] replyVars;`
instead.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 29]
```diff
@@ -0,0 +1,254 @@
...
+
+	LOG.debug("XmpDetector: isServiceDetected starting to query "+address);
+	
+        // try to establish session
+        aSession = new XmpSession(sockopts,address,xmpPort,xmpAuthenUser);
+        if (aSession == null) {
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: aSession cannot be null here. Please remove unnecessary if.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 30]
```diff
@@ -0,0 +1,254 @@
...
+	    new XmpVar("core","xmpdVersion","","",Xmp.SYNTAX_NULLSYNTAX),
+        };
+      
+        if ((aReply = aSession.queryVars(vars)) == null) {
+            LOG.debug("XmpDetector: isServiceDetected no vars from "+address);
+	    aSession.closeSession();
+            return false;
+        }
+
+        aSession.closeSession();
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: if You do not want to use multiple `aSession.closeSession` statements you can simply rewrite the method to something like this:

    Session aSession = initSession();
    try {
        // some code...
     }  finally {
        aSession.closeSession();
     }

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java [hunk 31]
```diff
@@ -0,0 +1,254 @@
...
+            return false;
+
+        } /* if replyVars == null */
+
+        LOG.debug("XmpDetector: isServiceDetected "+address+" reports "+
+                   replyVars[0].getValue()+","+
+                   replyVars[1].getValue());
+        LOG.debug("XmpDetector: isServiceDetected true for "+address);
+
+        LOG.debug("XmpDetector: isServiceDetected true for "+address);
```
Comment (mvrueden) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Please do not log the same message twice.

**Comment by mvrueden** [comment]
Hello @rdkgit, 

I looked at your code and added a couple of notes.

Here are a couple of considerations as well:
 * The code should be formatted as the Java defaults. Basically that means the { is at the end of the line.
 * The code should never contain System.out.printlns
 * The code does not need to worry about to check if LOG is null.
 * The code should not contain unnecessary reformattings. One can prevent those by simply invoke `git add -i -p` to manually and interactively add changes to your commit. 
  * The code should not contain out-commented lines.
  * Comments should not describe obvious things, here are some bad examples:

            // iterate over children
            for (Child eachChild : parent.getChildren()) {
            }

            // initialize a
            int a = 5;

**Comment by rdkgit** [comment]
You guys are killing me with the nitpicking!  :)  But seriously, have you guys looked at some of the code in ONMS lately?

**Comment by rdkgit** [comment]
Are you guys going to pull the XmpDetector in or not?  It wasnt clear from your messages whether you were merely critiquing my artistry or whether there were syntactic or architectural issues with my submission.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 22 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: This is really a silly critique.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 25 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Got it.  Honestly, it really doesn't matter much especially when you consider the state of some of the other code I've read over and copied.  Will fix.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 24 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Actually, I'd like to keep it as part of the comment since it helps me remember all the wonderful time I spent tracking down the bug.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 26 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Got it.  Fixed.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 27 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Its necessary because I'm setting up SSL socket stuff and I have to catch certificate issues.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 28 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: I'm going to assume this comment is a joke.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 29 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: True.  Thanks.  However, I still have to make the point that given some of the other ONMS code I've reviewed, I'm struggling to understand why you would hold up this submission over a extraneous if statement.

Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Actually, I need to check to see if aSession state = CLOSED.  Thanks for catching this.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 31 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Got it.  Again, I think this is nitpicking.  I've removed the duplicate.

Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Made the changes and am test compiling the code.  Do you need me to issue another pull request?

Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Made a bunch of the changes and pushed them (I think) upstream.  Let me know if you'd like me to create another pull request.

Thanks,

Bobby

**Comment by fooker** [comment]
Ther should be no need for another pull request. Just commit and push your changes to the same branche thid pull request is based on and it will be updated automatically. So we can discuss, fix an re-discuss the code until it's ready to merge.

Don't take the comments from marcus personally ;). As you alresdy mentioned, some (most?) parts of the code are not in a good shape, too. Thats one of the reasons why we are glad to have him in back in our team. It's just bad luck you are the first one who gets hit by his pernickety. Hopefully others will follow and we can engance the code quallity by this.

File: protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java (hunk 31 above)
Comment (rdkgit) [inline protocols/xmp/src/main/java/org/opennms/netmgt/protocols/xmp/detector/XmpDetector.java]: Actually, sorry, but git snafu.  I forgot to "commit".

I've now committed changes to XmpDetector, xmp-datacollection-config.xml and pushed them back upstream.  Please do let me know if I need to make other changes before you merge.  Thanks.

**Comment by rdkgit** [comment]
Hi!

>    Ther should be no need for another pull request. Just commit and push
>    your changes to the same branche thid pull request is based on and it
>    will be updated automatically. So we can discuss, fix an re-discuss the
>    code until it's ready to merge.

Just pushed up changes to two files (based on Markus' comments and a
bug fix I just found/applied to xmp-datacollection-config.xml).

Thanks,

Bobby

>    Don't take the comments from marcus personally ;). As you alresdy
>    mentioned, some (most?) parts of the code are not in a good shape, too.
>    Thats one of the reasons why we are glad to have him in back in our
>    team. It's just bad luck you are the first one who gets hit by his
>    pernickety. Hopefully others will follow and we can engance the code
>    quallity by this.

(I'm not.  However, coding conventions do not necessarily imply better
code nor documentation . . . .)

**Comment by mvrueden** [comment]
This pull request is closed because it originates from the wrong branch.
I created a new branch "jira/NMS-7404" and manually merged the changes from rdkgits branch "XmpMonitor" in.
A new Pull Request is also created for this purpose: https://github.com/OpenNMS/opennms/pull/282
